cmd/customd/client/client.go

```diff
@@ -0,0 +1,448 @@
+// and caches recent nonce locally to quickly sign
+type Nonce struct {
+	mutex     sync.Mutex
+	client    Client
+	addr      weave.Address
+	nonce     int64
+	fromQuery bool
+}
+
+// NewNonce creates a nonce for a client / address pair.
+// Call Query to force a query, Next to use cache if possible
+func NewNonce(client Client, addr weave.Address) *Nonce {
+	return &Nonce{client: client, addr: addr}
+}
+
+// Query always queries the blockchain for the next nonce
+func (n *Nonce) Query() (int64, error) {
+	user, err := n.client.GetUser(n.addr)
+	if err != nil {
+		return 0, err
+	}
+	n.mutex.Lock()
+	if user != nil {
+		n.nonce = user.UserData.Sequence
+	} else {
+		n.nonce = 0 // new account starts at 0
+	}
+	n.fromQuery = true
+	n.mutex.Unlock()
+	return n.nonce, nil
+}
+
+// Next will use a cached value if present, otherwise Query
+// It will always increment by 1, assuming last nonce
+// was properly used. This is designed for cases where
+// you want to rapidly generate many tranasactions without
+// querying the blockchain each time
+func (n *Nonce) Next() (int64, error) {
+	n.mutex.Lock()
+	initializedFromBlockchain := !n.fromQuery && n.nonce == 0
+	n.mutex.Unlock()
+	if initializedFromBlockchain {
+		return n.Query()
+	}
+	n.mutex.Lock()
+	n.nonce++
+	n.fromQuery = false
+	result := n.nonce
+	n.mutex.Unlock()
+	return result, nil
+}
+
+//************ generic (weave) functionality *************//
+
+// Status will return the raw status from the node
+func (cc *CustomClient) Status() (*Status, error) {
+	return cc.conn.Status()
+}
+
+// Genesis will return the genesis directly from the node
+func (cc *CustomClient) Genesis() (*GenesisDoc, error) {
+	gen, err := cc.conn.Genesis()
+	if err != nil {
+		return nil, err
+	}
+	return gen.Genesis, nil
+}
+
+// ChainID will parse out the chainID from the status result
+func (cc *CustomClient) ChainID() (string, error) {
+	gen, err := cc.Genesis()
+	if err != nil {
+		return "", err
+	}
+	return gen.ChainID, nil
+}
+
+// Height will parse out the Height from the status result
+func (cc *CustomClient) Height() (int64, error) {
+	status, err := cc.conn.Status()
+	if err != nil {
+		return -1, err
+	}
+	return status.SyncInfo.LatestBlockHeight, nil
+}
+
+// AbciResponse contains a query result:
+// a (possibly empty) list of key-value pairs, and the height
+// at which it queried
+type AbciResponse struct {
+	// a list of key/value pairs
+	Models []weave.Model
+	Height int64
+}
+
+// AbciQuery calls abci query on tendermint rpc,
+// verifies if it is an error or empty, and if there is
+// data pulls out the ResultSets from keys and values into
+// a useful AbciResponse struct
+func (cc *CustomClient) AbciQuery(path string, data []byte) (AbciResponse, error) {
+	var out AbciResponse
+
+	q, err := cc.conn.ABCIQuery(path, data)
+	if err != nil {
+		return out, err
+	}
+	resp := q.Response
+	if resp.IsErr() {
+		return out, errors.Errorf("(%d): %s", resp.Code, resp.Log)
```

Comment:
Please use `weave/errors.Wrapf` instead.

Reply:
In order to use `wrapf`, first paramater must be error. Should I use existing errors from `errors` package or create in the client?

Reply:
There is `ABCIError` i see now :+1: